Add Django middleware

jjjordanmsft opened the pull request:
This adds a Django middleware that loads telemetry client configuration from the application's `settings.py` file.  In addition to being simpler to configure, it will:

* Properly log exceptions and parent them to the requests.
* Derive request and operation names to the view that is serving the request (+ the http method).
* Annotate the request object with the name of the template that was rendered.
* Add an `appinsights` property to requests that are passed to views.  This can be used to generate telemetry that will then be parented to the request (think log messages and events).
* Disable telemetry or send to a separate ikey automatically when the Django variable `DEBUG=True`.

This also adds a `LoggingHandler` that reads the same metadata as the middleware, and can be included via Django's `LOGGING` variable.  The documentation covers this and short examples can be found in the tests.

This also addresses a few problems I came across while working on it:
* Sphinx documentation hit an infinite recursion due to the toctrees, so I sorted that out.
* The LogHandler did not set the severity level on the telemetry items.  Python's logging levels exactly match the AI schema's, so that can be 1-1 (fixed).
* I added `NullSender` because adding null checks for scenarios where I don't want telemetry drives me nuts.  All SDKs should have it.

msftclas commented:
@jjjordanmsft,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

AlexBulankou commented on README.md:
```diff
@@ -187,3 +187,109 @@ def hello_world():
+    
+    # (optional) By default, if DEBUG is True, then the middleware will
+    # not log any events.  To override this behavior, set debug_ikey, below:
+    'debug_ikey': "00000000-0000-0000-0000-000000000000",
```
What if I explicitly use telemetry client and generate some telemetry? Will the same rules with regards to DEBUG also apply? Or this only applies to telemetry auto-collected by django? I think if we want this feature it shouldn't be part of enablement for django, it should be part of sender

jjjordanmsft replied:
No.  `settings.py` and its `DEBUG` variable are Django-specific, and I intentionally avoided any Django dependencies outside of the `django` module.  The `appinsights.client` property found on views will use these same rules but if you're creating a TelemetryClient anywhere else, then it wouldn't.  I *do* have a factory method that could be exposed (it's technically already exposed, but not advertised) to get the same function.

AlexBulankou commented on README.md:
```diff
@@ -187,3 +187,109 @@ def hello_world():
+    
+    # (optional) By default, request names are logged as the fully-qualified
+    # name of the view.  To disable this behavior, specify:
+    'use_operation_url': True,
```
I think we should be consistent with ASP.NET and Node.JS and log "GET /Index" or Request URL as full URL

OsvaldoRosado replied:
+1

jjjordanmsft replied:
Fair enough, I'll make this change.

AlexBulankou commented on README.md:
```diff
@@ -187,3 +187,109 @@ def hello_world():
+    
+    # (optional) By default, arguments to views are tracked as custom
+    # properties.  To disable this, specify:
+    'record_view_arguments': False,
```
I think False should be default for consistency with other SDKs

jjjordanmsft replied:
Goes hand-in-hand with the above- if your operations are view names then you'd want to be able to segment and filter on the values.  I'll make it off-by-default.

AlexBulankou commented on README.md:
```diff
@@ -187,3 +187,109 @@ def hello_world():
+    # send_time specifies how long the sender waits for new input before recycling
+    # the background thread.
+    'send_interval': 1.0, # Check every second
+    'send_time': 3.0, # Wait up to 3 seconds for an event
```
Same question as above - do these apply to auto-collected events for django only or to any telemetry client that I create? If the former, I think they should not be, if the latter they should not be in django section and instead in Transmission Behavior section

jjjordanmsft replied:
Same answer as above- this would introduce Django dependencies outside of the `django` module, but we could supply factory methods.

AlexBulankou commented on README.md:
````diff
@@ -187,3 +187,109 @@ def hello_world():
     app.run()
 ```
+
+**Integrating with Django**
````
Thank you JJ!!!

AlexBulankou commented on applicationinsights/django/middleware.py:
```diff
@@ -0,0 +1,266 @@
+        # Fill in data from the response
+        data.duration = addon.measure_duration()
+        data.response_code = response.status_code
+        data.success = response.status_code < 400
```
should be 400 or >=402. 401 is still considered success

jjjordanmsft replied:
Ok.

OsvaldoRosado commented on applicationinsights/channel/contracts/MessageData.py:
```diff
@@ -9,6 +10,19 @@ class MessageData(object):
     
     DATA_TYPE_NAME = 'MessageData'
     
+    PYTHON_LOGGING_LEVELS = {
```
Contracts might not be the right spot for this dictionary. If we follow the other SDKs, eventually these files will all be autogenerated from our schema

jjjordanmsft replied:
Could you suggest a better place?

OsvaldoRosado replied:
Perhaps we can make a constants file in contracts or channel. The primary thing is to get it out of any of the actual type files in contracts. In node.js we'll soon have /contracts/constants.ts and then /contracts/generated/*.ts as we're switching that over to fully autogenerated schema-based types (like this MessageData one). Placing this in its own file, even if it stays in the same folder will be good enough while none of it is autogenerated yet here

OsvaldoRosado commented on applicationinsights/django/common.py:
```diff
@@ -0,0 +1,93 @@
+    "endpoint"])
+
+def load_settings():
+    if hasattr(settings, "APPLICATION_INSIGHTS"):
```
Just curious. Why both with and without underscore?

jjjordanmsft replied:
Because some people -- like me -- are inattentive and forgetful :-)

OsvaldoRosado commented on applicationinsights/django/common.py:
```diff
@@ -0,0 +1,93 @@
+        channel = saved_channels[channel_settings]
+    else:
+        if channel_settings.endpoint is not None:
+            sender = applicationinsights.channel.AsynchronousSender(service_endpoint_uri=channel_settings.endpoint)
```
Is this if necessary? Would setting the variable to None be the same as not setting it anyway?

jjjordanmsft replied:
This is coming off of the tuple, from line 33:

```
        endpoint = config.get("endpoint")
```

which will return `None` if it is not specified.

OsvaldoRosado replied:
Right, what I'm saying is if it is none the line will still work as it will just run
`applicationinsights.channel.AsynchronousSender(service_endpoint_uri=None)`

jjjordanmsft replied:
None is different from not setting the parameter.  I'd then be overriding the default value from [AsynchronousSender's constructor](https://github.com/Microsoft/ApplicationInsights-Python/blob/master/applicationinsights/channel/AsynchronousSender.py#L16)

OsvaldoRosado replied:
Ah got it. Strange that endpoint is defined there rather than a central config.. Oh well :)

OsvaldoRosado commented on applicationinsights/django/middleware.py:
```diff
@@ -0,0 +1,266 @@
+            
+            # (optional, uncommon) If you must send to an endpoint other than the
+            # default endpoint, specify it here:
+            'endpoint': "https://dc.services.visualstudio.com/v2/track",
```
Perhaps this is a good time to switch to https://dc.applicationinsights.microsoft.com @AlexBulankou?

ftxrc replied:
Would this enable live data?

OsvaldoRosado replied:
This change wouldn't affect any SDK functionality, it's just a different domain that points to the same place

OsvaldoRosado commented on applicationinsights/django/middleware.py:
```diff
@@ -0,0 +1,266 @@
+
+        return None
+
+    def process_exception(self, request, exception):
```
Node.js allows users to configure if they want automatic request/exceptions/dependencies as separate options. Maybe we should allow that here? (in case you want requests but no exceptions, etc)

jjjordanmsft replied:
Ok.

jjjordanmsft replied:
Addressed in update.

OsvaldoRosado commented on applicationinsights/django/middleware.py:
```diff
@@ -0,0 +1,266 @@
+        if hasattr(request, 'appinsights'):
+            client.context.operation.parent_id = request.appinsights.request.id
+
+        client.track_exception(type(exception), exception, sys.exc_info()[2])
```
Is the indexer here safe? Should there be a length & null check first?

jjjordanmsft replied:
Generally, yes:

```
Python 2.7.13 (default, Dec 21 2016, 07:16:46)
[GCC 6.2.1 20160830] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys; print(repr(sys.exc_info()))
(None, None, None)
```

```
Python 3.6.0 (default, Jan 16 2017, 12:12:55)
[GCC 6.3.1 20170109] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys; print(repr(sys.exc_info()))
(None, None, None)
>>>
```

But it couldn't hurt.

OsvaldoRosado commented on django_tests/all_tests.sh:
```diff
@@ -0,0 +1,56 @@
+#!/bin/bash
```
Is this tied into our Travis CI tests?

jjjordanmsft replied:
Not yet, and this is something I wanted consultation on.  Travis CI can make a matrix for you and test each combination, if you want.  e.g.:

```
# Tell Travis you want a Python environment to test in
language: python
# List the versions of Python you'd like to test against
python:
  - "2.5"
  - "2.6"
  - "2.7"
# List the versions of Django you'd like to test against
env:
  - DJANGO_VERSION=1.0.4
  - DJANGO_VERSION=1.1.4
  - DJANGO_VERSION=1.2.7
  - DJANGO_VERSION=1.3.1
  - DJANGO_VERSION=1.4
```

(lifted from [here](http://datadesk.latimes.com/posts/2012/06/test-your-django-app-with-travisci/))

But the matrix is trickier- Django 1.7 doesn't work with Python 3.5+, for example.  I could use the matrix and check for compatibility before running the test in the bash part.  Or, we could just stick with the supported Python versions and run all_tests, which will install and test each supported version of Django.

OsvaldoRosado replied:
Running all_tests makes sense to me if that's how it's expected to work locally anyway. Just realized there's no .bat equivalent in the PR for Windows 😛

jjjordanmsft replied:
I suppose I could whip something up.

OsvaldoRosado replied:
Did you have a chance to tie in all_tests to the travis configuration?

OsvaldoRosado commented on setup.py:
```diff
@@ -14,7 +14,7 @@
     # the version across setup.py and the project code, see
     # http://packaging.python.org/en/latest/tutorial.html#version
-    version='0.10.0',
+    version='0.11.0',
```
We should probably update the version number only when we're just about to merge develop to master

jjjordanmsft replied:
Ok.

ftxrc commented:
This'd be awesome to see working. Can't wait for it to be merged :+1:

jjjordanmsft commented:
I think this ticks all the boxes.  I've just removed 'debug_*' settings, exposed a factory method, and added the tests to Travis.  I haven't made the refactor to move the log level but it sounds to me like you're going to reorganize some things anyway.
